internal/proxyservice/proxy_service.go

```diff
@@ -57,78 +59,246 @@ func (s *proxyServices) Stop() error {
+		tcp:  tcpConfigs,
+		http: httpServiceConfigs,
+	}
+}
 
-		// TODO: Deal with SSH in a hardcoded way
+func (s *proxyServices) servicesToStart() (servicesToStart []internal.Service) {
+	configsByType := newConfigsByType(s.config.Services, s.availPlugins)
+	httpPlugins := s.availPlugins.HTTPPlugins()
+	tcpPlugins := s.availPlugins.TCPPlugins()
 
-		// Default case: not found
-		// TODO: Add Fatalf to our logger and use that
-		s.logger.Panicf("plugin '%s' not available.", requestedPlugin)
+	// TCP Plugins
+	for _, cfg := range configsByType.tcp {
+		// Validation will have already happened
+		tcpSvc, err := s.createTCPService(cfg, tcpPlugins[cfg.Connector])
+		if err != nil {
+			// TODO: Add Fatalf to our logger and use that
+			s.logger.Panicf("unable to create TCP service '%s'", cfg.Name)
+		}
+		servicesToStart = append(servicesToStart, tcpSvc)
+	}
+
+	// HTTP Plugins
+	for _, httpSvcConfig := range configsByType.http {
+		// Validation will have already happened
+		httpSvc, err := s.createHTTPService(
+			httpSvcConfig,
+			httpPlugins,
+		)
+		if err != nil {
+			// TODO: Add Fatalf to our logger and use that
+			s.logger.Panicf(
+				"unable to create HTTP proxy service on: '%s'",
+				httpSvcConfig.SharedListenOn,
+			)
+		}
+		servicesToStart = append(servicesToStart, httpSvc)
 	}
+
+	// TODO: Deal with SSH in a hardcoded way
+
 	return servicesToStart
 }
 
+func (s *proxyServices) createHTTPService(
+	httpSvcConfig HTTPServiceConfig,
+	plugins map[string]http.Plugin,
+) (internal.Service, error) {
+
+	// Create the listener
+
+	listener, err := net.Listen(
+		"tcp",
+		strings.TrimLeft(httpSvcConfig.SharedListenOn, "tcp://"),
+	)
+	if err != nil {
+		s.logger.Errorf("could not create listener on: %s", httpSvcConfig.SharedListenOn)
+		return nil, err
+	}
+
+	// Create the subservices
+
+	var subservices []httpproxy.Subservice
+	for _, subCfg := range httpSvcConfig.SubserviceConfigs {
+
+		// "cur" naming prefix needed to avoid package name collision
+		curPlugin := plugins[subCfg.Connector]
+		connResources := s.connectorResources(subCfg)
+		curConnector := curPlugin.NewConnector(connResources)
+
+		// Temp var required so that the function closes over the current loop
+		// value.
+		credsCopy := subCfg.Credentials
+		credsRetriever := func() (map[string][]byte, error) {
+			return GetSecrets(credsCopy)
+		}
+
+		subservices = append(subservices, httpproxy.Subservice{
+			Connector:           curConnector,
+			RetrieveCredentials: credsRetriever,
+		})
+	}
+
+	// Create the logger
+
+	// Name the http proxy service by its address
+	// (subservices have their own loggers)
+	proxyName := fmt.Sprintf("HTTP Proxy on %s", httpSvcConfig.SharedListenOn)
+	svcLogger := s.logger.CopyWith(proxyName, s.logger.DebugEnabled())
+
+	newSvc, err := httpproxy.NewProxyService(subservices, listener, svcLogger)
+	if err != nil {
+		s.logger.Errorf("could not create http proxy service '%s'", proxyName)
+		return nil, err
+	}
+	return newSvc, nil
+}
```

Comment:
I'm guessing that this method's implementation (or http-specific parts) might be a better fit for its own module/package to split concerns

Reply:
I'll think about this some more, but do keep in mind in that:
```go
	newSvc, err := httpproxy.NewProxyService(subservices, listener, svcLogger)
```
which is the punchline of this method is already split out, and that's where the http specific logic will this.  This method you're commenting on here is only responsible for putting the arguments together that get passed down to the real proxy service.  So it's fairly cohesive as is.  That said, it could probably be improved further.